what4/src/What4/Protocol/SMTLib2.hs

```diff
@@ -1166,6 +1176,24 @@ parseExpr sym sexp = case sexp of
     BVProof{} <- getBVProof arg1_expr
     BVProof{} <- getBVProof arg2_expr
     liftIO $ Some <$> I.bvConcat sym arg1_expr arg2_expr
+  SApp [SApp ["_", "extract", i_sexp, j_sexp], arg]
+    | Just i <- parseIntSolverValue i_sexp
+    , Just j <- parseIntSolverValue j_sexp -> do
+      let n = i - j + 1
+      Some j_repr <- return $ mkNatRepr $ fromIntegral j
+      Some n_repr <- return $ mkNatRepr $ fromIntegral n
+      Some arg_expr <- parseExpr sym arg
+      BVProof w_repr <- getBVProof arg_expr
+      case (isPosNat n_repr, testLeq (addNat j_repr n_repr) w_repr) of
+        (Just LeqProof, Just LeqProof) ->
+          liftIO $ Some <$> I.bvSelect sym j_repr n_repr arg_expr
+        _ -> throwError ""
+  SApp ((SAtom operator) : operands)
+    | Just op <- HashMap.lookup operator (opTable @sym)
+    , Just assoc <- opAssoc op
+    , length operands > 2 -> case assoc of
+      LeftAssoc -> parseExpr sym $ foldl' (\acc arg -> SApp [SAtom operator, acc, arg]) (head operands) (tail operands)
+      RightAssoc -> parseExpr sym $ foldr' (\arg acc -> SApp [SAtom operator, arg, acc]) (last operands) (init operands)
   SApp ((SAtom operator) : operands) -> case HashMap.lookup operator (opTable @sym) of
     Just (Op1 arg_types fn) -> do
       args <- mapM (parseExpr sym) operands
```

Comment:
Both of these cases overlap to some degree. To make this code easier to follow, would you be OK with refactoring it to look more like this?

```hs
  SApp ((SAtom operator) : operands) -> case HashMap.lookup operator (opTable @sym) of
    Just op <- HashMap.lookup operator (opTable @sym)
      | Just assoc <- opAssoc op
      , length operands > 2 -> case assoc of
        LeftAssoc -> parseExpr sym $ foldl' (\acc arg -> SApp [SAtom operator, acc, arg]) (head operands) (tail operands)
        RightAssoc -> parseExpr sym $ foldr' (\arg acc -> SApp [SAtom operator, arg, acc]) (last operands) (init operands)
    Just (Op1 arg_types fn) -> do
      ...
```